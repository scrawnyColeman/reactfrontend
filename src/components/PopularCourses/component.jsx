import React from 'react';
import styled from 'styled-components';
import { colours } from '../../constants/styles';
import Button from '../Button/component';
import LoadingSpinner from '../LoadingSpinner/component';
import { useHistory } from 'react-router-dom';
import { iconArray } from '../../images/imageLoader';

const Container = styled.div`
    text-align: left;
`;
export const StyledWrapper = styled.div`
    display: grid;
    grid-template-columns: 1fr min-content;
    margin: 0 8px;
`;
const StyledCoursesContainer = styled.div`
    height: 500px;
    overflow-y: scroll;
    border: 2px solid ${colours.primary}
    border-top: none;
`;

const StyledText = styled.div`
    padding-right: 12px;
    font-size: 1.25rem;
`;
const StyledImg = styled.img`
    margin: auto;
    width: 50px;
    object-fit: cover;
`;
const StyledHeading = styled.span`
    width: 100%;
    color: ${colours.primary};
    display: flex;
    margin: 0 auto;
    justify-content: space-between;
    font-size: 1.5rem;
`;
const StyledLabel = styled.div`
    text-align: left;
    padding: 10px 8px 12px;
    font-size: 2rem;
    border: 2px solid ${colours.primary};
`;

const StyledSpecialLabel = styled.span`
    color: ${colours.primary};
`;

const PopularCourses = ({ courses, isLoading }) => {
    const history = useHistory();
    let counter = 1;
    courses.sort((a, b) => b.activeUsers - a.activeUsers);

    return (
        <Container>
            <StyledLabel>Top 5 courses:</StyledLabel>
            <StyledCoursesContainer>
                {isLoading ? (
                    <LoadingSpinner />
                ) : (
                    courses.slice(0, 5).map((course) => {
                        const { description, id, language, title, activeUsers } = course;
                        const languageIcon = iconArray.find((icon) => icon.path === language.iconPath);
                        return (
                            <StyledWrapper key={id}>
                                <StyledText>
                                    <StyledHeading>
                                        <p> {`${counter++}: ${title} `}</p>
                                        <p>[{language.language.toUpperCase()}]</p>
                                    </StyledHeading>
                                    <p>
                                        {description.slice(0, 180)}
                                        {description.length > 180 && '...'}
                                    </p>
                                    <span>
                                        <p>
                                            <StyledSpecialLabel>{activeUsers}</StyledSpecialLabel>
                                            {activeUsers > 1 ? ` people have` : ` person has`} taken this course!
                                        </p>
                                        <Button
                                            onClick={() => history.push({ pathname: `/learn/${id}` })}
                                            text="Go"
                                            size="small"
                                            variant="outlined"
                                            hierarchy="primary"
                                        />
                                    </span>
                                </StyledText>
                                <StyledImg src={languageIcon.icon} alt={`${language.language}`} />
                            </StyledWrapper>
                        );
                    })
                )}
            </StyledCoursesContainer>
        </Container>
    );
};
export default PopularCourses;
