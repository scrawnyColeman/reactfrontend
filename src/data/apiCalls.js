import axios from 'axios';

const BASE_URL = 'http://localhost:8080/njoy';
const authToken = sessionStorage.getItem('authToken');
const headers = { headers: { authorization: authToken } };

/****** Course fetchrequests *******/
export const fetchAllLessons = () => axios.get(`${BASE_URL}/lessons`, headers);

export const fetchLessonById = (id) => axios.get(`${BASE_URL}/lessons/${id}`, headers);

export const fetchApprovedLessons = () => axios.get(`${BASE_URL}/lessons/approved`, headers);

export const fetchUnapprovedLessons = () => axios.get(`${BASE_URL}/lessons/unapproved`, headers);

export const fetchUserLessons = (id) => {
    return axios.get(`${BASE_URL}/user_lessons/${id}`, headers);
};

/****** Forum fetch requests *******/
export const fetchAllForumPosts = () => axios.get(`${BASE_URL}/forum_posts`, headers);

export const fetchForumPostByLesson = (id) => axios.get(`${BASE_URL}/forum_posts/lesson/${id}`, headers);

export const fetchForumPost = (id) => axios.get(`${BASE_URL}/forum_posts/${id}`, headers);

export const fetchForumPostComments = (id) => axios.get(`${BASE_URL}/comments/post/${id}`, headers);

/****** Practicals fetch requests *******/
export const fetchPracticalByLessonId = (id) => axios.get(`${BASE_URL}/practicals/lesson/${id}`, headers);

export const fetchSolutionByPracticalId = (id) => axios.get(`${BASE_URL}/solutions/practical/${id}`, headers);

/****** Quizzes fetch requests *******/
export const fetchAllQuestionsByLessonId = (id) => axios.get(`${BASE_URL}/questions/lesson/${id}`, headers);

export const fetchAllAnswersByLessonId = (id) => axios.get(`${BASE_URL}/questions/answers/lesson/${id}`, headers);

/****** Languages fetch requests *******/
export const fetchAllLanguages = () => axios.get(`${BASE_URL}/languages`, headers);

/****** Users fetch requests *******/
export const fetchUserByUsernameOrEmail = (username, email) =>
    axios.get(`${BASE_URL}/user/${username}/${email}`, headers);

/****** Users update requests *******/
export const updateUserInfo = (userId, username, email) =>
    axios.put(`${BASE_URL}/users/${userId}/username/${username}/email/${email}`, headers);

/****** Users post requests *******/
export const setUserLesson = (userId, lessonId) => axios.post(`${BASE_URL}/user/${userId}/lesson/${lessonId}`, headers);

export const createForumPost = (post) => axios.post(`${BASE_URL}/forum_posts`, post, headers);

export const createComment = (comment, userId, postId) =>
    axios.post(`${BASE_URL}/comments/${userId}/${postId}`, comment, headers);

export const deleteComment = (commentId) => axios.put(`${BASE_URL}/comments/delete/${commentId}`, headers);

export const editComment = (commentId, comment) =>
    axios.put(`${BASE_URL}/comments/edit/${commentId}`, comment, headers);

export const deleteLesson = (lessonId) => axios.delete(`${BASE_URL}/lessons/delete/${lessonId}`, headers);

export const approveLesson = (lessonId) => axios.put(`${BASE_URL}/lessons/approve/${lessonId}`, headers);

export const postLesson = (lesson, languageId) => axios.post(`${BASE_URL}/lessons/new/${languageId}`, lesson, headers);

export const postQuestions = (questions, lessonId) =>
    axios.post(`${BASE_URL}/lessons/questions/new/${lessonId}`, questions, headers);

export const postAnswers = (answers, questionId) =>
    axios.post(`${BASE_URL}/lessons/answers/new/${questionId}`, answers, headers);

export const postPractical = (practical, lessonId) =>
    axios.post(`${BASE_URL}/lessons/practical/new/${lessonId}`, practical, headers);

export const postSolution = (solution, practicalId) =>
    axios.post(`${BASE_URL}/lessons/solutions/new/${practicalId}`, solution, headers);

export const setLessonComplete = (lessonId, userId) =>
    axios.put(`${BASE_URL}/lesson/complete/${lessonId}/${userId}`, headers);
