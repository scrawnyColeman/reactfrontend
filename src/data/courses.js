import JavaIcon from '../images/java.png';
import JavaImg from '../images/javaheader.jpg';
import JSIcon from '../images/js.png';
import JSImg from '../images/jsimage.jpg';
import PythonIcon from '../images/python.png';
import PythonImg from '../images/python image.jpg';

const data = {
    courses: [
        {
            title: 'ReactJS',
            language: 'JavaScript',
            description:
                'Learn React, an open-source JavaScript library, for building fast webapps and clever User interfaces.',
            icon: JSIcon,
            header: JSImg,
            id: '1',
            activeUsers: '3514', //how many users have taken this course to generate popularity ranking
            currentUsers: ['1'],
            prereqs: [],
            theory: `yes`,
            youtubeLink: null,
        },
        {
            title: 'Lambda Functions',
            language: 'Java',
            description:
                'Produce concise lambda functions to enhance your functional programming skills. It is recommended that users complete our interfaces lesson first.',
            icon: JavaIcon,
            header: JavaImg,
            id: '2',
            activeUsers: '38',
            currentUsers: ['1', '2'],
            prereqs: ['1', '2'],
            theory: `yes`,
            youtubeLink: null,
        },
        {
            title: 'Recursion',
            language: 'JavaScript',
            description:
                'Learn recursive functions to ace all of your coding interviews. We recommend that all users complete our loops and conditionals lesson before beginning here.',
            icon: JSIcon,
            header: JSImg,
            id: '3',
            activeUsers: '802',
            currentUsers: ['0'],
            prereqs: ['1', '2'],
            theory: `A common computer programming tactic is to divide a problem into sub-problems of the same type as the original, solve those sub-problems, and combine the results. This is often referred to as the divide-and-conquer method; when combined with a lookup table that stores the results of solving sub-problems (to avoid solving them repeatedly and incurring extra computation time), it can be referred to as dynamic programming or memoization.\n
            A recursive function definition has one or more base cases, meaning input(s) for which the function produces a result trivially (without recurring), and one or more recursive cases, meaning input(s) for which the program recurs (calls itself). For example, the factorial function can be defined recursively by the equations 0! = 1 and, for all n > 0, n! = n(n − 1)!. Neither equation by itself constitutes a complete definition; the first is the base case, and the second is the recursive case. Because the base case breaks the chain of recursion, it is sometimes also called the "terminating case".\n
            The job of the recursive cases can be seen as breaking down complex inputs into simpler ones. In a properly designed recursive function, with each recursive call, the input problem must be simplified in such a way that eventually the base case must be reached. (Functions that are not intended to terminate under normal circumstances—for example, some system and server processes—are an exception to this.) Neglecting to write a base case, or testing for it incorrectly, can cause an infinite loop.\n
            For some functions (such as one that computes the series for e = 1/0! + 1/1! + 1/2! + 1/3! + ...) there is not an obvious base case implied by the input data; for these one may add a parameter (such as the number of terms to be added, in our series example) to provide a 'stopping criterion' that establishes the base case. Such an example is more naturally treated by co-recursion, where successive terms in the output are the partial sums; this can be converted to a recursion by using the indexing parameter to say "compute the nth term (nth partial sum)" `,
            youtubeLink: `https://www.youtube.com/watch?v=Mv9NEXX1VHc`,
        },
        {
            title: 'Polymorphism',
            language: 'Java',
            description:
                'Meaning to take many forms, polymorphism is among the four pillars of Object-oriented programming; a must know for all programmers.',
            icon: JavaIcon,
            header: JavaImg,
            id: '4',
            activeUsers: '604',
            currentUsers: ['1', '2'],
            prereqs: ['1', '2'],
            theory: `The word polymorphism is used in various contexts and describes situations in which something occurs in several different forms. In computer science, it describes the concept that objects of different types can be accessed through the same interface. Each type can provide its own, independent implementation of this interface. It is one of the core concepts of object-oriented programming (OOP). \nIf you’re wondering if an object is polymorphic, you can perform a simple test. If the object successfully passes multiple is-a or instanceof tests, it’s polymorphic. As I’ve described in my post about inheritance, all Java classes extend the class Object. Due to this, all objects in Java are polymorphic because they pass at least two instanceof checks.`,
            youtubeLink: `https://www.youtube.com/watch?v=tIWm3I_Zu7I`,
        },
        {
            title: 'Control Flow and Conditionals',
            language: 'Python',
            description:
                'Learn how to write a program that is able to do hundreds of hours of human cognition in seconds through repetition and scripted decision making',
            icon: PythonIcon,
            header: PythonImg,
            id: '5',
            activeUsers: '190',
            currentUsers: ['0'],
            prereqs: ['1', '2'],
            theory: `yes`,
            youtubeLink: null,
        },
        {
            title: 'Classes and Objects',
            language: 'Python',
            description: 'Learn the fundamental elements that make up the world of Object-Oriented Programming.',
            icon: PythonIcon,
            header: PythonImg,
            id: '6',
            activeUsers: '334',
            currentUsers: ['0'],
            prereqs: ['1', '2'],
            theory: `yes`,
            youtubeLink: null,
        },
        {
            title: 'ES6 Arrow Functions',
            language: 'JavaScript',
            description:
                'Learn arrow functions - introduced with the ECMAScript 6 - and discover how to write shorter, more concise functions with more capable than before.',
            icon: JSIcon,
            header: JSImg,
            id: '7',
            activeUsers: '2154',
            currentUsers: ['0'],
            prereqs: ['1', '2'],
            theory: `yes`,
            youtubeLink: null,
        },
    ],
};

export default data;
