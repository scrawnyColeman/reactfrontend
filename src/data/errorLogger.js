export const errorLogger = (err) => {
    console.log(err);
};
