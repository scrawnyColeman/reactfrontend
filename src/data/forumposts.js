import axios from 'axios';

const BASE_URL = 'http://localhost:8080/njoy';

export const fetchAllForumPosts = () => axios.get(`${BASE_URL}/forum_posts`);
export const fetchForumPostByLesson = (id) => axios.get(`${BASE_URL}/forum_posts/lesson/${id}`);
export const fetchForumPost = (id) => axios.get(`${BASE_URL}/forum_posts/${id}`);
export const fetchForumPostComments = (id) => axios.get(`${BASE_URL}/comments/post/${id}`);

export default fetchAllForumPosts;
