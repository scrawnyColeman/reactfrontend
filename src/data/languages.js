export const data = {
    languages: [
        {
            id: '1',
            name: 'Java',
        },
        {
            id: '2',
            name: 'JavaScript',
        },
        {
            id: '3',
            name: 'Python',
        },
    ],
};
