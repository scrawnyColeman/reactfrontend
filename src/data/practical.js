import axios from 'axios';

const BASE_URL = `http://localhost:8080/njoy`;
export const fetchPracticalByLessonId = (id) => axios.get(`${BASE_URL}/practicals/lesson/${id}`);
export const fetchSolutionByPracticalId = (id) => axios.get(`${BASE_URL}/solutions/practical/${id}`);

export default fetchPracticalByLessonId;
