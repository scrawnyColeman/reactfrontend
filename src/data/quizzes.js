import axios from 'axios';

const BASE_URL = `http://localhost:8080/njoy`;

export const fetchAllQuestionsByLessonId = (id) => axios.get(`${BASE_URL}/questions/lesson/${id}`);
export const fetchAllAnswersByLessonId = (id) => axios.get(`${BASE_URL}/questions/answers/lesson/${id}`);

export default fetchAllAnswersByLessonId;
