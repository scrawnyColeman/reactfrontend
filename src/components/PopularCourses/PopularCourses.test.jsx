import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-styled-components';
import React from 'react';
import PopularCourses from './component';
import data from '../../data/courses.js'

Enzyme.configure({ adapter: new Adapter() });

describe('PopularCourses component', () => {
    test('component matches snapshot', () => {
        const wrapper = shallow(<PopularCourses courses={data.courses} />);
        expect(wrapper).toMatchSnapshot();
    });
});
