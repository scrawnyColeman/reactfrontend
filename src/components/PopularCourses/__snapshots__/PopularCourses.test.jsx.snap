// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularCourses component component matches snapshot when loading 1`] = `
<styled.div>
  <styled.div>
    Top 5 courses:
  </styled.div>
  <styled.div>
    <LoadingSpinner />
  </styled.div>
</styled.div>
`;

exports[`PopularCourses component component matches snapshot when not oading 1`] = `
<styled.div>
  <styled.div>
    Top 5 courses:
  </styled.div>
  <styled.div>
    <styled.div
      key="5"
    >
      <styled.div>
        <styled.span>
          <p>
             
            1: Loops 
          </p>
          <p>
            [
            JAVA
            ]
          </p>
        </styled.span>
        <p>
          Learn how to utilise Loops in your programs.
        </p>
        <span>
          <p>
            This course has been selected 
            <styled.span>
              1
            </styled.span>
             time(s).
          </p>
          <Button
            hierarchy="primary"
            onClick={[Function]}
            size="small"
            text="Go"
            variant="outlined"
          />
        </span>
      </styled.div>
      <styled.img
        alt="java"
        src="java.png"
      />
    </styled.div>
  </styled.div>
</styled.div>
`;
